### PR TITLE
Remove pseudo-commands for mid-execution capture.

### PR DESCRIPTION
**Pull request by AWoloszyn**
Instead of serializing pseudo-commands for mid-execution capture "eg: RecreateInstance", serialize the state block as-is.

Only re-create the state block when we need to on the server side.

TODO in a future MR:
 - Remove the cases where we generate the commands when not necessary
  - 1) Generating the dependency graph
  - 2) Resolving resources

### CODE REVIEW

File: gapii/cc/spy_base.h [hunk 1]
```diff
@@ -48,6 +48,7 @@
 #include <unordered_map>
 
 namespace gapii {
+const uint8_t kAllAPIs = 0xFF;
```
Comment (Qining) [inline gapii/cc/spy_base.h]: Probably we can change the `getEncoder` to make is accept a bit mask, instead the shift value, so we can avoid using this 'mask' like thing with the 'shift value' like thing in should_trace()? Hmm, probably not in this CL anyway.

File: gapis/api/vulkan/command_buffer_rebuilder.go [hunk 2]
```diff
@@ -927,13 +1069,12 @@ func rebuildVkCmdDebugMarkerInsertEXT(
 			markerNameData.Free()
 			markerInfoData.Free()
 		}, cb.VkCmdDebugMarkerInsertEXT(commandBuffer, markerInfoData.Ptr()).AddRead(
-			markerNameData.Data()).AddRead(markerInfoData.Data())
+			markerNameData.Data()).AddRead(markerInfoData.Data()), nil
 }
 
 func GetCommandArgs(ctx context.Context,
```
Comment (Qining) [inline gapis/api/vulkan/command_buffer_rebuilder.go]: Probably we should add comment for it?

File: gapis/api/vulkan/externs.go [hunk 3]
```diff
@@ -79,6 +80,10 @@ func (e externs) callSub(ctx context.Context, cmd api.Cmd, id api.CmdID, s *api.
 	})
 }
 
+func (e externs) callSub(ctx context.Context, cmd api.Cmd, id api.CmdID, s *api.GlobalState, b *rb.Builder, sub, data interface{}) {
```
Comment (Qining) [inline gapis/api/vulkan/externs.go]: Drop it? It is only used once.

File: gapis/api/vulkan/command_buffer_rebuilder.go [hunk 4]
```diff
@@ -1035,6 +1176,108 @@ func GetCommandArgs(ctx context.Context,
 	}
 }
 
+func GetCommandFunction(cr CommandReference) interface{} {
```
Comment (Qining) [inline gapis/api/vulkan/command_buffer_rebuilder.go]: blurbs?

File: gapis/api/vulkan/state_rebuilder.go [hunk 5]
```diff
@@ -0,0 +1,2582 @@
...
+	newState      *api.GlobalState
+	cmds          []api.Cmd
+	cb            CommandBuilder
+	readMemories  []*api.AllocResult
+	writeMemories []*api.AllocResult
+}
+
+// TODO: wherever possible, use old resources instead of doing full reads on the old pools.
+//       This is especially useful for things that are internal pools, (Shader words for example)
+func (s *State) RebuildState(ctx context.Context, oldState *api.GlobalState) []api.Cmd {
```
Comment (Qining) [inline gapis/api/vulkan/state_rebuilder.go]: Blurbs?

File: gapis/api/vulkan/state_rebuilder.go [hunk 6]
```diff
@@ -0,0 +1,2582 @@
...
+			if handled {
+				handledPipelines[k] = true
+				delete(unhandledPipelines, k)
+				numHandled++
+			}
+		}
+		if numHandled == 0 {
+			// There is a cycle in the basePipeline indices.
+			// Or the no base pipelines does exist.
+			// Create the rest without base pipelines
```
Comment (Qining) [inline gapis/api/vulkan/state_rebuilder.go]: Is it allowed to by cyclical? Probably we should emit an error in the report view for this?

File: gapis/api/vulkan/state_rebuilder.go [hunk 7]
```diff
@@ -0,0 +1,2582 @@
...
+		}
+	} else {
+		for _, p := range s.GraphicsPipelines.KeysSorted() {
+			pp := s.GraphicsPipelines.Get(p)
+			unhandledPipelines[pp.VulkanHandle] = pp.BasePipeline
+		}
+	}
+
+	numHandled := 0
+	for len(unhandledPipelines) != 0 {
```
Comment (Qining) [inline gapis/api/vulkan/state_rebuilder.go]: Looks like not optimal? Anyway, we can optimize it later if it matters.

File: gapis/api/vulkan/state_rebuilder.go (hunk 6 above)
Comment (AWoloszyn) [reply gapis/api/vulkan/state_rebuilder.go]: I am not sure if it is technically allowed, but I want to at least continue to work if it is.
Hopefully the validation layers can catch this.

File: gapis/api/vulkan/state_rebuilder.go [hunk 8]
```diff
@@ -0,0 +1,2582 @@
...
+
+	// Make sure we allocate a buffer that is more than big enough for the
+	// data
+	sb.write(sb.cb.VkAllocateMemory(
+		device.VulkanHandle,
+		NewVkMemoryAllocateInfoᶜᵖ(sb.MustAllocReadData(
+			VkMemoryAllocateInfo{
+				VkStructureType_VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO,
+				NewVoidᶜᵖ(memory.Nullptr),
+				size * 2, // Overallocate by a factor of 2.
```
[the remark this answers is not in the document]
Comment (AWoloszyn) [reply gapis/api/vulkan/state_rebuilder.go]: Discussed offline.

File: gapis/api/vulkan/state_rebuilder.go [hunk 9]
```diff
@@ -0,0 +1,2568 @@
...
+				handledPipelines[k] = true
+				delete(unhandledPipelines, k)
+				numHandled++
+			}
+		}
+		if numHandled == 0 {
+			// There is a cycle in the basePipeline indices.
+			// Or the no base pipelines does exist.
+			// Create the rest without base pipelines
+			for k, _ := range unhandledPipelines {
```
Comment (ben-clayton) [inline gapis/api/vulkan/state_rebuilder.go]: Superdooperübernit:

```go
for k := range unhandledPipelines {
```

Is more idiomatic

File: gapis/api/vulkan/state_rebuilder.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ben-clayton) [reply gapis/api/vulkan/state_rebuilder.go]: Would be nice to be documented in the code. 😉

File: gapis/api/vulkan/state_rebuilder.go [hunk 10]
```diff
@@ -0,0 +1,2592 @@
...
+		1,
+		sb.MustAllocReadData(VkMappedMemoryRange{
+			VkStructureType_VK_STRUCTURE_TYPE_MAPPED_MEMORY_RANGE,
+			NewVoidᶜᵖ(memory.Nullptr),
+			deviceMemory,
+			VkDeviceSize(0),
+			size,
+		}).Ptr(),
+		VkResult_VK_SUCCESS,
+	).AddWrite(dat.Data()))
```
[the remark this answers is not in the document]
Comment (ben-clayton) [reply gapis/api/vulkan/state_rebuilder.go]: Ping

File: gapis/api/vulkan/state_rebuilder.go [hunk 11]
```diff
@@ -0,0 +1,2582 @@
...
+
+func (sb *stateBuilder) freeScratchBuffer(device *DeviceObject, buffer VkBuffer, mem VkDeviceMemory) {
+	sb.write(sb.cb.VkDestroyBuffer(device.VulkanHandle, buffer, memory.Nullptr))
+	sb.write(sb.cb.VkFreeMemory(device.VulkanHandle, mem, memory.Nullptr))
+}
+
+func (sb *stateBuilder) getSparseQueueFor(lastBoundQueue *QueueObject, device VkDevice, queueFamilyIndices *map[uint32]uint32) *QueueObject {
+	hasQueueFamilyIndices := queueFamilyIndices != nil
+
+	queueProperties := sb.s.PhysicalDevices.Get(sb.s.Devices.Get(lastBoundQueue.Device).PhysicalDevice).QueueFamilyProperties
```
[the remark this answers is not in the document]
Comment (ben-clayton) [reply gapis/api/vulkan/state_rebuilder.go]: Ping

File: gapis/api/vulkan/vulkan.go [hunk 12]
```diff
@@ -74,9 +75,19 @@ func (*State) Root(ctx context.Context, p *path.State) (path.Node, error) {
 	return p, nil
 }
 
-func (*State) SetupInitialState(ctx context.Context) {}
-
-func (*State) RebuildState(ctx context.Context, s *api.GlobalState) []api.Cmd { return nil }
+func (c *State) SetupInitialState(ctx context.Context) {
```
Comment (ben-clayton) [inline gapis/api/vulkan/vulkan.go]: Docs please

File: gapii/cc/slice.h [hunk 13]
```diff
@@ -90,7 +90,10 @@ inline Slice<T>::Slice() : mBase(nullptr), mCount(0) {}
 template<typename T>
 inline Slice<T>::Slice(T* base, uint64_t count, const std::shared_ptr<Pool>& pool)
     : mBase(base), mCount(count), mPool(pool) {
-    GAPID_ASSERT(mBase != nullptr || count == 0 /* Slice: null pointer */);
+
+    if (!mPool || !mPool->is_virtual()) {
```
[the remark this answers is not in the document]
Comment (AWoloszyn) [reply gapii/cc/slice.h]: IIRC mPool is nullptr, if this is the application pool. I might be wrong, but I think that is the reasoning for this logic.

File: gapii/cc/spy.cpp [hunk 14]
```diff
@@ -483,14 +510,14 @@ void Spy::onPostFrameBoundary(bool isStartOfFrame) {
             exit();
             set_suspended(false);
             saveInitialSate();
-            set_recording_state(true);
+            //set_recording_state(true);
```
[the remark this answers is not in the document]
Comment (AWoloszyn) [reply gapii/cc/spy.cpp]: That should all be cleaned up and gone now. Not sure why it shows in this CR.

File: gapii/cc/spy_base.h (hunk 1 above)
Comment (AWoloszyn) [reply gapii/cc/spy_base.h]: Yeah, I would prefer to leave this for a future CL.

File: gapii/cc/chunk_writer.cpp [hunk 15]
```diff
@@ -69,7 +69,9 @@ bool ChunkWriterImpl::write(std::string& s) {
 }
 
 void ChunkWriterImpl::flush() {
-    mStreamGood = mWriter->write(mBuffer.data(), mBuffer.size());
+    size_t bufferSize = mBuffer.size();
+    // If we flush en empty buffer, this will return false, which is incorrect.
```
[the remark this answers is not in the document]
Comment (AWoloszyn) [reply gapii/cc/chunk_writer.cpp]: That is correct, removed the comment.

File: gapii/cc/pool.h [hunk 16]
```diff
@@ -25,6 +25,9 @@ namespace gapii {
 class Pool {
 public:
     static std::shared_ptr<Pool> create(uint32_t id, uint64_t size);
+// This creates a pool that can be serialized, but has no actual
```
[the remark this answers is not in the document]
Comment (AWoloszyn) [reply gapii/cc/pool.h]: done

**Pull request merged**